types.go

```diff
@@ -236,6 +206,14 @@ type Settings struct {
 	Recorder Recorder
 }
 
+// TODO Comment
```

Comment:
Do we really have to decide the full Problem struct now? There is no code that needs constraints now and apparently it is unclear how the nonlinear constraints should be represented.

How about having two functions, one for equality and one for inequality constraints, that return a slice of functions? Similarly for their gradient, but then it gets a bit more complicated. I am not sure, but having an access to individual constraints functions could be useful. I don't understand your remark about calling Constraints with nil arguments the first time, but whatever it means it doesn't sound like something fancy.

I would say that Hess et al. should take matrix interfaces, it would leave the door open for any future implementation of sparse problems when there are sparse symmetric matrices, for example. Unless there are some drawbacks that I'm not aware of.